### DISCUSSION TITLE
`MCUboot` cannot boot the main app on `SAMD21`

### DISCUSSION THREAD

**Discussion by vshymanskyy** [zephyrproject-rtos/zephyr#67220] [Q&A] [answered]
I'm trying to use `adafruit_feather_m0_basic_proto` board with MCUboot.

I've added `bootloader/mcuboot/boot/zephyr/boards/adafruit_feather_m0_basic_proto.overlay`:

```dts
/ {
    chosen {
        zephyr,code-partition = &slot0_partition;
    };
};

/* Remove default ATSAMD21 partitions */
/delete-node/ &boot_partition;
/delete-node/ &code_partition;
/delete-node/ &storage_partition;

&flash0 {
    partitions {
        compatible = "fixed-partitions";
        #address-cells = <1>;
        #size-cells = <1>;

        boot0_partition: partition@0 {
            label = "sam-ba";
            reg = <0x0 DT_SIZE_K(8)>;
            read-only;
        };
        boot_partition: partition@2000 {
            label = "mcuboot";
            reg = <0x0002000 DT_SIZE_K(32)>;
            read-only;
        };
        slot0_partition: partition@A000 {
            label = "image-0";
            reg = <0x000A000 DT_SIZE_K(92)>;
        };
        slot1_partition: partition@21000 {
            label = "image-1";
            reg = <0x0021000 DT_SIZE_K(92)>;
        };
        nvs_partition: partition@38000 {
            label = "nvs";
            reg = <0x0038000 DT_SIZE_K(24)>;
        };
        scratch_partition: partition@3E000 {
            label = "scratch";
            reg = <0x003E000 DT_SIZE_K(8)>;
        };
    };
};

```

I've also added in `zephyr/samples/application_development/sysbuild/with_mcuboot/sysbuild/mcuboot.conf`:
```
CONFIG_BOOT_MAX_IMG_SECTORS=384
```

Then, the sample project is built:
```sh
west build -p -b adafruit_feather_m0_basic_proto --sysbuild samples/application_development/sysbuild/with_mcuboot
```

Flashing is performed in 2 steps:
```sh
# Flash bootloader (double-click the Reset button to enter the SAM Boot Assistant)
west -v flash --build-dir build/mcuboot
# Flash slot0 (double-click the Reset button to enter the SAM Boot Assistant)
west -v flash --build-dir build/with_mcuboot
```

As a result:
```log
I: Starting bootloader
I: Primary image: magic=unset, swap_type=0x1, copy_done=0x3, image_ok=0x3
I: Scratch: magic=unset, swap_type=0x1, copy_done=0x3, image_ok=0x3
I: Boot source: primary slot
I: Image index: 0, Swap type: none
I: Bootloader chainload address offset: 0xa000
I: Jumping to the first image slot
```
And nothing more is printed (the board freezes?)

Is it a bug, a non-feature, or maybe I've missed something?

**Comment by thedjnK** [comment]
Works fine here on nordic platforms so something wrong with the atmel code/hal/driver/etc. @nandojve @attie-argentum @pdgendt @mnkp

**Comment by vshymanskyy** [comment]
Yes, it works on many other boards here as well ;)

**Reply by pdgendt** [reply]
I'm not familiar with the Cortex-M0 variants of the Atmel SAM controllers, my guess would be that this issue is similar to the Cortex-M4 code I had, where the clock initialization can only be done once.
With both the bootloader and the application code trying to setup some clocks, this isn't supported.

I have an ongoing PR for some of the Cortex-M4 variants here: https://github.com/zephyrproject-rtos/zephyr/pull/66499. But again, I don't know the `sam0` code.

**Reply by nandojve** [reply]
There is no official support to MCUboot on any Atmel boards in Zephyr as far I know. The SAME/V (SAM) works for sure because I helped to support on NuttX in the past but I never made it available officially on Zephyr. In the case of SAMD21, which is a SAM0, I never tested on this type of processor. MCUboot is not a plug-in-play SW and there are few things that need to be well aligned to make it work. This may includes Zephyr Kconfig and all sign process.

What officially is supported for SAM is SAM-BA and for SAM0 SAM-BA in UF2 or Arduino mode.

The board you mention is already using SAM-BA as bootloader. This means that SAM-BA will boot MCUBoot which should boot the Zephyr App. I'm not sure if this is the recommended way to do it, besides it could work with more effort. Since this is a very small board I recommend you to remove the SAM-BA and put MCUboot in the first partition. However, if you really want to have MCUboot as second stage bootloader you need to dig how configure to build the App with right flags and do the proper sign. I assume that you need a bundle image with correct offsets to flash it on HW or you need to use a commercial tool like J-Link. Don't forget to align the partitions on a way that respects the flash sectors.

I strong believe that it is possible to use MCUboot with SAMD because I think someone from community is using with SAME54. This means that the flash driver already supports all requirements for MCUboot

**Reply by vshymanskyy** [reply]
Yes, SAM-BA loads the MCUboot successfully (as per the configuration above), and MCUboot detects the `slot0` firmware successfully.
It locks up when transferring control to the main firmware, so it looks like there's a minor disconnection here.

**Reply by ndrs-pst** [reply]
@vshymanskyy, To avoid a deadlock in GCLK0, you might need to redirect GCLK0 from DFLL48M to a different clock source, such as GCLK_GENCTRL_SRC_GCLKGEN1. This can be achieved by inserting the following code:
```
GCLK->GENCTRL.reg = GCLK_GENCTRL_ID(0)
			  | GCLK_GENCTRL_SRC_GCLKGEN1
			  | GCLK_GENCTRL_IDC
			  | GCLK_GENCTRL_GENEN;
```
This adjustment should be made at [this section of the code](https://github.com/zephyrproject-rtos/zephyr/blob/main/soc/arm/atmel_sam0/common/soc_samd2x.c#L161C1-L161C1). The necessity arises because z_arm_platform_init() is executed twice: first by MCUBoot, where GCLK0 uses DFLL48M, and then by the main firmware. If GCLK0 isn't rerouted, the code at [these lines](https://github.com/zephyrproject-rtos/zephyr/blob/main/soc/arm/atmel_sam0/common/soc_samd2x.c#L168-L170) will disable DFLL48M, causing a deadlock in GCLK0.

**Reply by nandojve** [reply, marked as answer]
Hi @vshymanskyy ,

I tested and confirm that @ndrs-pst is right. There is an issue when switching clocks. I send the #67275 to fix it and I could use MCuboot on atsamr21_xpro (formally a samd21 + rf233).

MCuboot Patch:

```diff
diff --git a/boot/zephyr/boards/atsamr21_xpro.conf b/boot/zephyr/boards/atsamr21_xpro.conf
new file mode 100644
index 00000000..71261122
--- /dev/null
+++ b/boot/zephyr/boards/atsamr21_xpro.conf
@@ -0,0 +1,6 @@
+# Copyright 2023 Gerson Fernando Budke <nandojve@gmail.com>
+# SPDX-License-Identifier: Apache-2.0
+
+CONFIG_BOOT_MAX_IMG_SECTORS=512
+CONFIG_BOOT_UPGRADE_ONLY=y
+CONFIG_MCUBOOT_DOWNGRADE_PREVENTION=y

```

Zephyr Patch:

```diff
diff --git a/boards/arm/atsamr21_xpro/atsamr21_xpro.dts b/boards/arm/atsamr21_xpro/atsamr21_xpro.dts
index 86981d86cf..0ca453c0af 100644
--- a/boards/arm/atsamr21_xpro/atsamr21_xpro.dts
+++ b/boards/arm/atsamr21_xpro/atsamr21_xpro.dts
@@ -19,6 +19,7 @@
                zephyr,shell-uart = &sercom0;
                zephyr,sram = &sram0;
                zephyr,flash = &flash0;
+               zephyr,code-partition = &slot0_partition;
                zephyr,ieee802154 = &ieee802154;
        };
 
@@ -214,14 +215,22 @@ ext2_i2c: &sercom1 {
                #address-cells = <1>;
                #size-cells = <1>;
 
-               /*
-                * The final 16 KiB is reserved for the application.
-                * Storage partition will be used by FCB/LittleFS/NVS
-                * if enabled.
-                */
-               storage_partition: partition@3c000 {
-                       label = "storage";
-                       reg = <0x0003c000 0x00004000>;
+               boot_partition: partition@0 {
+                       label = "mcuboot";
+                       reg = <0x00000000 DT_SIZE_K(64)>;
+                       read-only;
+               };
+               slot0_partition: partition@10000 {
+                       label = "image-0";
+                       reg = <0x00010000 DT_SIZE_K(80)>;
+               };
+               slot1_partition: partition@24000 {
+                       label = "image-1";
+                       reg = <0x00024000 DT_SIZE_K(80)>;
+               };
+               scratch_partition: partition@38000 {
+                       label = "scratch";
+                       reg = <0x00038000 DT_SIZE_K(2)>;
                };
        };
 };
diff --git a/boards/arm/atsamr21_xpro/doc/index.rst b/boards/arm/atsamr21_xpro/doc/index.rst
index da9cb14a43..857d07f89d 100644
--- a/boards/arm/atsamr21_xpro/doc/index.rst
+++ b/boards/arm/atsamr21_xpro/doc/index.rst
@@ -210,6 +210,16 @@ Flashing
 
    You should see "Hello World! atsamr21_xpro" in your terminal.
 
+MCUboot build
+=============
+
+west build -b atsamr21_xpro -d build/mcuboot-atsamr21_xpro bootloader/mcuboot/boot/zephyr
+west flash -d build/mcuboot-atsamr21_xpro
+
+west build -b atsamr21_xpro -d build/app zephyr/samples/hello_world
+west sign -t imgtool -d build/app -p `pwd`/bootloader/mcuboot/scripts/imgtool.py -- --version 1.0.0 --pad --key bootloader/mcuboot/root-rsa-2048.pem
+west flash -d build/app --hex-file build/app/zephyr/zephyr.signed.hex
+
 References
 **********
 
diff --git a/dts/arm/atmel/samd2x.dtsi b/dts/arm/atmel/samd2x.dtsi
index c247daffc1..dc514f13a8 100644
--- a/dts/arm/atmel/samd2x.dtsi
+++ b/dts/arm/atmel/samd2x.dtsi
@@ -71,6 +71,7 @@
                        flash0: flash@0 {
                                compatible = "soc-nv-flash";
                                write-block-size = <4>;
+                               erase-block-size = <64>;
                        };
                };
 
```

I used a different partition from you but this shouldn't be a problem. I noted that CONFIG_BOOT_UPGRADE_ONLY=y boots the application. The BOOT_SWAP_USING_SCRATCH=y do not boot and I did not dig on it. I didn't test an upgrade to make sure flash driver is fine.

I hope you can move forward with this. If you find anything else open a PR and share with us!

**Reply by vshymanskyy** [reply]
@nandojve I can confirm that https://github.com/zephyrproject-rtos/zephyr/pull/67275 fixes the issue.
Thank you very much!

**Reply by henrikbrixandersen** [reply]
https://github.com/zephyrproject-rtos/zephyr/pull/67275 is now merged into `main`.

**Comment by thaobk91** [comment]
Hi @vshymanskyy ,
I am developing MCUboot for samd21g18j
Could you share me a instruction for mcuboot with that MCU?
